siege/siege.py

```diff
@@ -0,0 +1,194 @@
+  # e.g. with a command-line option.
+  reps = 2000
+  siege_args = ["--reps=%d" % reps, "--rc=" + siege_dir + "/siege.conf", 
+                proxy_url]
+
+  # Clean up any old log files, so the CSVs just show the results from
+  # the runs we are about to do.
+  for fname in csv_files + [log]:
+    if os.path.exists(fname):
+      os.remove(fname)
+
+  with open(log, "a") as logfile:
+    # Echoes a command to the logfile, and then returns it. This is
+    # intended to wrap the command argment to subprocess.Popen and
+    # subprocess.call so we can see in the log how they were run.
+    def echo(command):
+      logfile.write(" ".join(command) + "\n")
+      logfile.flush()
+      return command
+
+    # Starts up Envoy that serves as both an origin and a proxy, and then
+    # sieges it with ~50k requests, measuring the overall throughput, errors,
+    # and memory usage. The path to the binary and filenames to collect
+    # performance CSV into a file that's passed in, and return some memory info
+    # as a triple.
+    def runEnvoyAndSiege(envoy_binary, csv_files):
+      # Run Envoy in the background and wait for it to respond on the upstream
+      # port, admin port, and proxy port.
+      envoy = subprocess.Popen(echo([envoy_binary, "-c", envoy_conf]),
+                               stdout=logfile, stderr=logfile)
+
+      for url in [admin, proxy_url, upstream_url]:
+        waitForUrl(url)
+
+      # Siege the envoy.
+      subprocess.call(echo(["siege", "--log=" + csv_files[0]] + siege_args),
+                      stdout=logfile, stderr=logfile)
+
+      # Capture Envoy's opinion of how much memory it's using, and also
+      # ask the OS what it thinks via ps. Append those to a separate CSV
+      # file from the one harvested by siege.
+      statmem = loadJson(admin + "/memory")["allocated"]
+
+      vsz = 0
+      rsz = 0
+      pid_str = "%d" % envoy.pid
+      for line in runCapturingStdout("ps -eo pid,vsz,rsz").split("\n"):
+        pid_vsz_rsz = line.split(" ")
+        if pid_vsz_rsz[0] == pid_str:
+          vsz = pid_vsz_rsz[1]
+          rsz = pid_vsz_rsz[2]
+          break
+
+      with open(csv_files[1], "a") as mem_csv:
+        mem_csv.write("%s,%s,%s\n" % (statmem, vsz, rsz))
+
+      # Send Envoy a quit request and wait for the process to exit.
+      requests.post(admin + "/quitquitquit")
+      envoy.wait()
+
+    for mem_csv in [clean_mem_csv, experimental_mem_csv]:
+      with open(mem_csv, "w") as mem_csv_file:
+        mem_csv_file.write("EnvoyMem, VSZ, RSS\n")
+
+    # Loops 5x interleaving runs with the clean and experimental versions
+    # of Envoy, collecting the results in separate CSV files.
+    progress("Logging 10 runs to %s " % log)
+    for run in range(0, 4):
+      runEnvoyAndSiege(clean_envoy, clean_csv_files)
+      progress("...%d" % (2 * run + 1))
+      runEnvoyAndSiege(experimental_envoy, experimental_csv_files)
+      progress("...%d" % (2 * run + 2))
+
+    print("\n")
```

Comment:
Not a super huge fan of verbose plain text based output unless guarded by something like `-d` for debug. Ideally these are tools we are building for automation, so we should think in terms of what would be machine consumed.